enhancements/network/egress-qos.md

```diff
@@ -0,0 +1,107 @@
+---
+title: OVN Pods Egress DSCP QoS
+authors:
+  - "@oribon"
+reviewers:
+  - "@trozet"
+approvers:
+  - TBD
+creation-date: 2022-02-16
+last-updated: 2022-02-16
+status: implementable
+---
+
+# OVN Pods Egress DSCP QoS
+
+## Summary
+
+Not all traffic has the same priority, and when there is contention for bandwidth, there should be a mechanism for objects outside the cluster to prioritize the traffic.
+To enable this, we will use Differentiated Services Code Point (DSCP) which allows us to classify packets by setting a 6-bit field in the IP header, effectively marking the priority of a given packet relative to other packets as "Critical", "High Priority", "Best Effort" and so on.
+
+By introducing a new CRD `PodQos`, users could specify a DSCP value for packets originating from pods on a set of namespaces heading to a specified CIDR.
+The CRs will be watched by ovn-k, which in turn will configure OVN's [QoS Table](https://man7.org/linux/man-pages/man5/ovn-nb.5.html#QoS_TABLE). 
+
+## Motivation
+
+Telco customers require support for DSCP marking capability for some of their 5G applications, giving some pods precedence over others.
+The QoS markings will be consumed and acted upon by objects outside of the OpenShift cluster to optimize traffic flow throughout their networks.
+
+### Goals
+
+- Provide a mechanism for users to set DSCP on egress traffic coming from specific namespaces.
+
+### Non-Goals
+
+- Ingress QoS.
+
+- Consolidating with current `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` annotations.
+Nonetheless, the work done here does not interfere with the current bandwidth QoS mechanism. # Q: Is this point relevant?
+
+- The DSCP marking does not need to be handled or acted upon by OpenShift, just added to selected headers.
+
+- Marking East/West traffic, exposing the DSCP value from the inner packet to the outer geneve packet. # Q: is this a goal? If so does it mean we allow dstCIDR to be empty / expect the user to specify an internal pod ip? https://www.mail-archive.com/ovs-dev@openvswitch.org/msg59787.html
+
+## Proposal
+
+To achieve egress DSCP marking on pods, we introduce a new CRD `PodQos`, which lets users specify a namespaceSelector, DSCP value and destination CIDR - meaning traffic coming from pods on specific namespaces heading to the destination CIDR will be marked with the given DSCP value.
```

Comment:
Why per-namespace? Why not per-pod? Why allow a single PodQoS object to apply to multiple namespaces, rather than requiring a separate PodQoS for each namespace? Why per-destination-CIDR rather than per-destination-port or based on time of day or anything else?

Even if the current decisions solve the problem for the current customer, will they solve the problem for all customers?

Reply:
I went with per-namespace and not per-pod because that's the current requirements (which tbh are pretty vague right now), while per-pod(s) could be added later on.
Rethinking with your comments, we could have the CRD namespace scoped, with subsections corresponding to a CIDR/+port, leaving room for further granularity like podSelectors in the future, while enforcing 1 PodQos per namespace.
That way it might be less ambiguous in comparison to a cluster scoped CRD - it'll be clearer from where each Namespace gets its QoS.
Something like:
```yaml
kind: PodQos
apiVersion: k8s.ovn.org/v1
metadata:
  name: default
  namespace: ns1
spec:
  egress:
    - dstCIDR: 0.0.0.0/0
      port: 80
      dscp: 46
      priority: 1 # needed if overlapping destinations are used?
      # PodSelector could be another field here in the future?
    - dstCIDR: 1.2.3.4/32
      port: 443
      dscp: 32
      priority: 2
...
```
Is this approach better/cleaner?

IMO using dstCIDR/+port (or none) to determine when to apply DSCP makes sense, and tbh I'm not too familiar with other possible requirements to tell if there's anything better as a starting point or if it contradicts with another option.

Reply:
> Like, if this object is _only_ going to be used to describe QoS-related stuff that is _only_ acted on by out-of-cluster components, then that suggests one design. But if we wanted to handle QoS-related stuff more generally, then that might suggest a different design, etc.

Well, I think our main (or maybe only) focus is the former. Going by that I updated the file to reflect that, pretty much following the `EgressFirewall` approach - having a namespace-scoped CRD `EgressQos`, affecting all of the pods in the resource's namespace. I think it makes sense in this context, and keeps this egress only - leaving room for maybe small additions like marking by destination port if we'd like.